src/writer.rs

```diff
@@ -0,0 +1,200 @@
+        } else {
+            #[cfg(feature = "logging")]
+            log::trace!("pre-pending {} bits", self.leftover.len());
+
+            self.leftover.extend_from_bitslice(bits);
+            &mut self.leftover
+        };
+
+        // one shot impl of BitSlice::read(no read_exact), but for no_std
+        let mut buf = alloc::vec![0x00; bits.len() / 8];
+        let mut count = 0;
+        bits.chunks_exact(bits_of::<u8>())
+            .zip(buf.iter_mut())
+            .for_each(|(byte, slot)| {
+                *slot = byte.load_be();
+                count += 1;
+            });
+
+        // SAFETY: This does noto have a safety comment in bitvec. But this is safe
+        // because of `count` here will always still be within the bounds
+        // of `bits`
+        bits = unsafe { bits.get_unchecked(count * bits_of::<u8>()..) };
+
+        // TODO: with_capacity?
+        self.leftover = bits.to_bitvec();
+        if self.inner.write_all(&buf).is_err() {
+            return Err(DekuError::Write);
+        }
+
+        self.bits_written += buf.len() * 8;
+        #[cfg(feature = "logging")]
+        log::trace!("wrote {} bits: {buf:02x?}", buf.len() * 8);
+
+        Ok(())
+    }
+
+    /// Write `buf` into `Writer`
+    ///
+    /// If no `self.leftover`, this will write directly into `Writer`, and if not will write
+    /// `buf` using `Self::write_bits()`.
+    // The following inline(always) helps performance significantly
+    #[inline(always)]
+    pub fn write_bytes(&mut self, buf: &[u8]) -> Result<(), DekuError> {
+        #[cfg(feature = "logging")]
+        log::trace!("writing {} bytes: {buf:02x?}", buf.len());
+
+        if !self.leftover.is_empty() {
+            #[cfg(feature = "logging")]
+            log::trace!("leftover exists");
+
+            // TODO: we could check here and only send the required bits to finish the byte?
+            // (instead of sending the entire thing)
+            self.write_bits(&BitVec::from_slice(buf))?;
+        } else {
+            if self.inner.write_all(buf).is_err() {
+                return Err(DekuError::Write);
+            }
+            self.bits_written += buf.len() * 8;
+        }
+
+        Ok(())
+    }
+
+    /// Write all remaining bits into `Writer`, adding empty bits to the end so that we can write
+    /// into a byte buffer
+    #[inline]
+    pub fn finalize(&mut self) -> Result<(), DekuError> {
+        if !self.leftover.is_empty() {
+            #[cfg(feature = "logging")]
+            log::trace!("finalized: {} bits leftover", self.leftover.len());
+
+            // add bits to be byte aligned so we can write
+            self.leftover
+                .extend_from_bitslice(&bitvec![u8, Msb0; 0; 8 - self.leftover.len()]);
+            let mut buf = alloc::vec![0x00; self.leftover.len() / 8];
+
+            // write as many leftover to the buffer (as we can, can't write bits just bytes)
+            // TODO: error if bits are leftover? (not bytes aligned)
```

Comment:
is this covered by some test case? I'm curious what it currently looks like when/if we hit this

Reply:
This is actually can't be hit, as we fill out that buffer to the length of a byte above.